Refactor schema, add Person schema

jdevalk opened the pull request:
## Summary

This PR can be summarized in the following changelog entry:

* Refactored the schema code, changing it all to one big Graph.
* Add Person markup for author pages.
* WebPage markup for all pages.
* Article markup for posts, with Person markup for the author.
* Adds MySpace, SoundCloud, Tumblr and YouTube URL input fields to people's profiles.
* Changes the "Company or Person" section of the Knowledge graph settings to allow selecting a an author that is the "Person" so we can do all the proper Person markup everywhere.

## Relevant technical choices:

* Refactored the code so that each "bit" of schema has its own underlying class, which implements an interface `WPSEO_Graph_Piece`. 
* The JSON+LD class has been replaced by `WPSEO_Schema`.

## To do
- [x] When `person` is selected in Company or Person, disable SEO -> Social -> Accounts
- [x] When `person` is selected in Company or Person, use Person's twitter account for twitter card output
- [x] When `person` is selected in Company or Person, but no actual Person is selected, throw an error.
- [ ] Update config wizard step.

## Test instructions
This PR can be tested by following these steps:

* Make sure _all_ the Schema markup is still there, although all of them are now in one Graph. 
* Added Person markup on Author archive pages.

## UI changes
* [x] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [ ] I have added unittests to verify the code works as intended

Fixes #12461 
Fixes #11079
Fixes #8432 
Fixes #12473 
Fixes #12026 
Fixes #10532
Fixes #8161 
Fixes #12515
Fixes #12414

jdevalk commented:
Related roadmap issue: https://github.com/Yoast/roadmap/issues/76

inetbiz commented:
Make Person Image a wp_media() call. Requires certain ratio and pixel requirements to get into a google knowledge graph once claimed. https://support.google.com/knowledgepanel/answer/7534842?hl=en subsection featured image.

jdevalk commented:
@inetbiz nothing in that section relates to a website, they only talk about what you can suggest there directly? Also, if you search for `joost de valk` right now, you'll see they feature an entirely square image of me at the moment :)

inetbiz commented:
@jdevalk KG image comes from a variety of online images and verified KG. The question in that line of code displays on an article/blogposting. If the `"@type":"WebSite",` is set, it's not valid. Anything other than the home page is an `ItemPage, CollectionPage, et al` It's towards the bottom => **More specific Types**

That's all I'm asking.

inetbiz commented:
@jdevalk The more I think about it: It's something you should blog about or put in theme dev docs category. ProfilePage on the theme's author.php if exist. ItemPage if not exists. Right direction to go with that for author entity in content?

jdevalk commented:
@inetbiz we're working on something far more extensive than what you're thinking about :)

jdevalk commented:
@inetbiz sample markup for the frontpage:

```html
<script type='application/ld+json'>{"@context":"https://schema.org","@graph":[{"@type":"Person","@id":"http://one.wordpress.test/#person","name":"Joost de Valk","image":"http://1.gravatar.com/avatar/ad5dd285560221a7302608c74c690035?s=96&d=mm&r=g","sameAs":["https://www.facebook.com/jdevalk","https://www.linkedin.com/in/jdevalk","https://twitter.com/jdevalk"]},{"@type":"WebSite","@id":"http://one.wordpress.test/#website","url":"http://one.wordpress.test/","name":"one.wordpress.test","publisher":{"@id":"http://one.wordpress.test/#person"},"potentialAction":{"@type":"SearchAction","target":"http://one.wordpress.test/?s={search_term_string}","query-input":"required name=search_term_string"}},{"@type":"WebPage","@id":"http://one.wordpress.test/#webpage","url":"http://one.wordpress.test/","inLanguage":"en-US","name":"one.wordpress.test - Just another WordPress site","isPartOf":{"@id":"http://one.wordpress.test/#website"},"description":"Just another WordPress site"}]}</script>
```

For a search results page:

```html
<script type='application/ld+json'>{"@context":"https://schema.org","@graph":[{"@type":"Person","@id":"http://one.wordpress.test/#person","name":"Joost de Valk","image":"http://1.gravatar.com/avatar/ad5dd285560221a7302608c74c690035?s=96&d=mm&r=g","sameAs":["https://www.facebook.com/jdevalk","https://www.linkedin.com/in/jdevalk","https://twitter.com/jdevalk"]},{"@type":"WebSite","@id":"http://one.wordpress.test/#website","url":"http://one.wordpress.test/","name":"one.wordpress.test","publisher":{"@id":"http://one.wordpress.test/#person"},"potentialAction":{"@type":"SearchAction","target":"http://one.wordpress.test/?s={search_term_string}","query-input":"required name=search_term_string"}},{"@type":"SearchResultsPage","@id":"http://one.wordpress.test/search/hello/#webpage","url":"http://one.wordpress.test/search/hello/","inLanguage":"en-US","name":"You searched for hello - one.wordpress.test","isPartOf":{"@id":"http://one.wordpress.test/#website"},"breadcrumb":{"@id":"http://one.wordpress.test/search/hello/#breadcrumb"}},{"@type":"BreadcrumbList","@id":"http://one.wordpress.test/search/hello/#breadcrumb","itemListElement":[{"@type":"ListItem","position":1,"item":{"@type":"WebPage","@id":"http://one.wordpress.test/","name":"Home"}},{"@type":"ListItem","position":2,"item":{"@type":"WebPage","@id":"http://one.wordpress.test/search/hello/","name":"You searched for \u201chello\u201d"}}]}]}</script>
```

and for an article page:

```html
<script type='application/ld+json'>{"@context":"https://schema.org","@graph":[{"@type":"Person","@id":"http://one.wordpress.test/#person","name":"Joost de Valk","image":"http://1.gravatar.com/avatar/ad5dd285560221a7302608c74c690035?s=96&d=mm&r=g","sameAs":["https://www.facebook.com/jdevalk","https://www.linkedin.com/in/jdevalk","https://twitter.com/jdevalk"]},{"@type":"WebSite","@id":"http://one.wordpress.test/#website","url":"http://one.wordpress.test/","name":"one.wordpress.test","publisher":{"@id":"http://one.wordpress.test/#person"},"potentialAction":{"@type":"SearchAction","target":"http://one.wordpress.test/?s={search_term_string}","query-input":"required name=search_term_string"}},{"@type":"WebPage","@id":"http://one.wordpress.test/hello-world/#webpage","url":"http://one.wordpress.test/hello-world/","inLanguage":"en-US","name":"Hello world! - one.wordpress.test","isPartOf":{"@id":"http://one.wordpress.test/#website"},"breadcrumb":{"@id":"http://one.wordpress.test/hello-world/#breadcrumb"}},{"@type":"BreadcrumbList","@id":"http://one.wordpress.test/hello-world/#breadcrumb","itemListElement":[{"@type":"ListItem","position":1,"item":{"@type":"WebPage","@id":"http://one.wordpress.test/","name":"Home"}},{"@type":"ListItem","position":2,"item":{"@type":"WebPage","@id":"http://one.wordpress.test/hello-world/","name":"Hello world!"}}]}]}</script>
```

Still a work in progress but I think this is already quite neat.

inetbiz commented:
listitem 2 for a page should be an ItemPage type not another WebPage. http://one.wordpress.test/hello-world/

inetbiz commented:
@jdevalk actually it's gets way more complicated aye? What if they have contactpage type and the breadcrumb is trying to define that as a webpage? I know that breadcrumbs plays a role in related entities patent for google

atimmer commented on frontend/schema/class-schema.php:
```diff
@@ -0,0 +1,103 @@
+			return wp_json_encode( $data, JSON_UNESCAPED_SLASHES ); // phpcs:ignore PHPCompatibility.Constants.NewConstants.json_unescaped_slashesFound -- Version check present.
+		}
+
+		return wp_json_encode( $data );
```
For debugging I found it helpful to replace format_data with the following code:

```php
	public function format_data( $data ) {
		$flags = 0;
		if ( version_compare( PHP_VERSION, '5.4', '>=' ) ) {
			$flags = $flags | JSON_UNESCAPED_SLASHES;
		}
		if ( defined( 'WPSEO_DEBUG' ) && WPSEO_DEBUG ) {
			$flags = $flags | JSON_PRETTY_PRINT;
		}
		return wp_json_encode( $data, $flags );
	}
```

(From https://github.com/Yoast/wordpress-seo/blob/feature/schema-blocks-framework/frontend/class-json-ld.php#L210-L222)

That makes it easier to debug the schema, because you can actually see the structure when you are working on it. When `WPSEO_DEBUG` is on.

jdevalk commented on admin/views/tabs/metas/paper-content/general/knowledge-graph.php:
```diff
@@ -38,7 +41,31 @@
+		echo '</span>';
+		if ( WPSEO_Options::get( 'company_or_person_user_id', false ) ) {
+			$user = get_userdata( WPSEO_Options::get( 'company_or_person_user_id' ) );
+			printf( esc_html__( 'You have selected the user %1$s as the person this site represents. Their user profile information will now be used in search results.', 'wordpress-seo' ), '<strong>' . $user->display_name . '</strong>' );
```
We discussed that, it's too much data to show here.

jdevalk commented on frontend/schema/interface-wpseo-graph-piece.php:
```diff
@@ -0,0 +1,28 @@
+		 *
+		 * @return bool
+		 */
+		public function is_needed();
```
Renamed it liked ten times myself, it was at first called `should_add`.

jdevalk commented on frontend/schema/interface-wpseo-graph-piece.php:
```diff
@@ -0,0 +1,28 @@
+		 *
+		 * @return array|bool $graph A graph piece on success, false on failure.
+		 */
+		public function add_to_graph();
```
will rename to `generate()`

jdevalk commented on frontend/schema/class-schema-article.php:
```diff
@@ -0,0 +1,137 @@
+	 * @return array $data Article data.
+	 */
+	public function add_to_graph() {
+		$post          = get_post();
```
Hmm you'll have to explain this one in person to me I think, don't get it yet :)

jdevalk replied:
Hmm I do get it, but think it should then be true for _all_ of these... At which point maybe `WPSEO_Graph_Piece` should not be an interface but an abstract.

jdevalk commented on frontend/schema/class-schema-article.php:
```diff
@@ -0,0 +1,137 @@
+	 * @return string
+	 */
+	private function get_publisher_url() {
+		if ( WPSEO_Options::get( 'company_or_person', '' ) !== 'company' ) {
```
`WPSEO_Site` ? Can add, could also move the `WPSEO_Utils` `home_url` function I use left and right to that....

atimmer replied:
Yes, `WPSEO_Site`. (I'm already looking forward to namespaces)

afercia commented:
I'd like to recommend a few things related to UI and accessibility. Basing my comments on this screenshot (so this are just initial thoughts based on a prototype):

![Screen Shot 2019-03-21 at 14 33 48](https://user-images.githubusercontent.com/1682452/54771446-98959c80-4c05-11e9-9d79-37c880e7c6c2.png)

## Required

To my understanding, this is a choice that becomes _required_ after selecting "Person". If this is the case, even before displaying errors the select needs to be indicated as a required choice.

For input fields, one common practice is to use an asterisk and explain what the asterisk means at the beginning of the form: "Fields marked with an asterisk are required". Plus, an HTML attribute "required" or "aria-required". This is a select though, not sure this pattern is the best option.

Another option is to just use plain text: `{label of the select} (required)` plus attribute.

## Labelling

Not sure "Name" is a proper label. The real associated action is selecting a user. Not sure why the text "Select a user" is the first option within the select, I'd say that should be the label instead. 

## Input validation and error reporting

This is a bit tricky and related to #12173. This choice is both in the Settings pages (PHP) and in the Configuration Wizard (JS). Should a validation mechanism be implemented for both PHP and JS?

- the error message should be easily perceivable by users with accessibility needs, including screen reader users
- with PHP there's a page reload:
  - the presence of an error should be communicated at the top of the page and focus moved there 
  - another technique is to use the document title to warn users there's an error
  - the item in error should be clearly identified: color alone is not sufficient: `aria-invalid` on the select may help
  - error details can be placed close to the item in error, like in the screenshot
  - for vision impairments, the red border wouldn't be sufficient to identify the item in error but I guess the presence of the error details _close_ to the select can be considered sufficient
  - alternatively, form submission and page reload can be prevented when there's an error and in this case everything should be done via JS
- with JS there's no page reload:
  - the error message should be communicated also with an audible message via `speak()`
  - the item in error should get an `aria-invalid="true"` attribute

I do realise put this way this appears to be a lot of work 🙂 One of the points on #12173 is: considering we're moving away from PHP (at least for the UI) maybe an input validation / error reporting mechanism should be implemented only in JS? On the other hand, one more thing to consider is the Voluntary Product Accessibility Template (VPAT) for wpseo. Without a solid validation mechanism we should state the plugin "doesn't support" input validation.

inetbiz commented on frontend/schema/class-schema-person.php:
```diff
@@ -0,0 +1,223 @@
+		$data['image']  = array(
+			'@type'   => 'ImageObject',
+			'@id'     => WPSEO_Utils::get_home_url() . '#personlogo',
+			'url'     => get_avatar_url( $user_data->user_email ),
```
This is going to have to be wp_media() because Google requires a certain pixel width and ratio before it will accept as a featured image for the KG. See https://support.google.com/knowledgepanel/answer/7534842?hl=en and go down to featured image.

jdevalk replied:
That's not what that documentation says.

jdevalk commented on frontend/schema/class-schema-article.php:
```diff
@@ -0,0 +1,137 @@
+ */
+
+/**
+ * Class WPSEO_Schema_Article
```
yeah fixed.

atimmer commented on frontend/schema/class-schema-article.php:
```diff
@@ -66,8 +81,8 @@ public function generate() {
 	private function get_author_url() {
-		if ( WPSEO_Options::get( 'company_or_person', '' ) === 'person' ) {
-			return WPSEO_Utils::home_url() . '#person';
+		if ( $this->context->site_represents === 'person' ) {
```
I like how you've combined my `WPSEO_Site` suggestions with the context.

atimmer commented on frontend/schema/class-schema-context.php:
```diff
@@ -0,0 +1,114 @@
+	 * WPSEO_Schema_Context constructor.
+	 */
+	public function __construct() {
+		$this->breadcrumbs_enabled = current_theme_supports( 'yoast-seo-breadcrumbs' );
```
Small nitpick: Constructors are not supposed to have side-effects.

atimmer commented on frontend/schema/interface-wpseo-graph-piece.php:
```diff
@@ -10,6 +10,12 @@
+		 *
+		 * @param WPSEO_Schema_Context $context A value object with context variables.
+		 */
+		public function __construct( WPSEO_Schema_Context $context );
```
I would remove this line. Enforcing the constructor to have a certain interface is restrictive and usually doesn't add a lot of benefit. If a graph piece in the future needs an extra dependency (ver likely) or doesn't use the `$context` at all (rare, but possible) that is not possible.

atimmer commented on frontend/schema/interface-wpseo-graph-piece.php:
```diff
@@ -0,0 +1,36 @@
+		 *
+		 * @param WPSEO_Schema_Context $context A value object with context variables.
+		 */
+		public function __construct( WPSEO_Schema_Context $context );
```
```suggestion
```

atimmer commented:
The last thing I would change is putting these `#article`, `#person` and `#primaryimage` strings in class constants. This makes sure they are referred to the same everywhere and you cannot make typos when using one of those somewhere new.

igorschoester commented:
Acceptance 🏗 

Tested with @dariaknl !



* Testing a new post with `Google Structured Data Testing Tool` it gave the error: `A value for the image field is required.`. We think this is Google being silly, is this really the case?
* Adding multiple tags results in multiple `keywords` entries in the output. According to schema.org comma separation is expected for keywords.
* When selecting a person in the knowledge graph settings you have to pick a user or you get an error. You can still save this. This is not clear and should not be possible.
* Person profile does not have a Wikipedia entry. This might make some sense for the average user, but still...
* Knowledge graph is set as a person:
    * In the testing tool the `publisher` of the article is type `Thing`. In the JSON-LD this is only an @id field with the expected `http://local.wordpress.test/author/admin/#person`.
    * Testing tool shows the error: `The attribute publisher.itemtype has an invalid value.`.
* When writing a post as a person we get this JSON-LD output (see below). A couple of things:
    * The first Person in the graph has the @id of the site url + #person. This is unexpected, as the person is not the site admin.
    * There are 2 publisher sections, one of the WebSite (`http://local.wordpress.test/#person`) and one of the Article (`http://local.wordpress.test/author/luella-bernier/#person`). Where `luella-bernier` is the 'user_nicename', this is a database only value. It seems this is not accessible from UI.
* Bug: When the knowledge graph setting is set to a person. In a post the JSON-LD's website publisher is ALWAYS set to the author of the post.
    * This is the same on an author archive.
* For custom post types, the schema type of Article is expected. However, we got the WebPage type.

jdevalk commented:
> Testing a new post with Google Structured Data Testing Tool it gave the error: A value for the image field is required.. We think this is Google being silly, is this really the case?

Well, it's their requirement, not schema.org's requirement :) but I can't make up an image that isn't there.

> Adding multiple tags results in multiple keywords entries in the output. According to schema.org comma separation is expected for keywords.

Ah, that's actually easier, fixed in my last push.

> When selecting a person in the knowledge graph settings you have to pick a user or you get an error. You can still save this. This is not clear and should not be possible.

True, but not something I can easily fix, which is why I chose the current implementation.

> Person profile does not have a Wikipedia entry. This might make some sense for the average user, but still...

True, adding, fixed in my last push.

There are obviously some bugs in the Author vs "site person" thing, I'm working on that right now, changing the implementation slightly.

igorschoester commented:
> True, but not something I can easily fix, which is why I chose the current implementation.

Maybe default to the current site admin instead of the question option?

jdevalk commented:
> Bug: When the knowledge graph setting is set to a person. In a post the JSON-LD's website publisher is ALWAYS set to the author of the post.
> This is the same on an author archive.

I think this is fixed now. 

> For custom post types, the schema type of Article is expected. However, we got the WebPage type.

Hmm that's wrong in the scenario. We only output Article on `post` post types by default.

igorschoester commented:
Testing to be continued... 2 findings so far

Situation:
- Knowledge graph setting on company.
- New post from `admin`.

JSON-LD output:
- Article author includes the `name`. This is already added in the `Person` schema. Is this necessary?
- Article and WebSite both have a publisher, which is always the same as the knowledge graph setting. Is this on purpose?


Situation:
- Knowledge graph setting on person `admin`.
- New post from `admin`.

JSON-LD output:
- An unexpected person `@id` is included `http://local.wordpress.test/#author`. We think this should end with `#person` as the WebSite publisher is set on this `@id`.
- Article author has the expected `@id`: `http://local.wordpress.test/author/admin/#author`. Which links to another seperate Person schema.

jdevalk commented:
> Article author includes the name. This is already added in the Person schema. Is this necessary?

Yes.

> Article and WebSite both have a publisher, which is always the same as the knowledge graph setting. Is this on purpose?

Yeah `publisher` is a required attribute for `Article`, that's why it's there.

2nd situation: good catch, fixed the bug.

igorschoester commented:
> Yeah publisher is a required attribute for Article, that's why it's there.

On schema.org I do not see any required status. Via the `Google Structured Data Testing Tool` I see the warning if I remove the publisher.
However, there is also one for a warning for an image. Which we do not always have.
Is there a tool that shows this in an overview?

jdevalk commented:
> > Yeah publisher is a required attribute for Article, that's why it's there.
> 
> On schema.org I do not see any required status. Via the `Google Structured Data Testing Tool` I see the warning if I remove the publisher.
> However, there is also one for a warning for an image. Which we do not always have.
> Is there a tool that shows this in an overview?

Google requires it, which is annoying, for Article markup to do anything meaningful. See [their docs](https://developers.google.com/search/docs/data-types/article). Same goes for image in many cases, but we can't force an image, we _have_ the publisher, so this is more of a pragmatic choice :)

I think Google needs to change a few things in their Structured Data Testing Tool (SDTT), especially when they say something is required and it's not required by schema.org. I'll talk to them about that.

igorschoester commented:
Tested again with @dariaknl from the start.

For pages when the knowledge graph is on a different person than the author of the page. We do not get to see the author information anywhere. Just to confirm, is this on purpose?
On schema.org, the WebPage schema does have an author field.

For author archive page when the knowledge graph is on a different person than the archive is for. We do not get to see the WebSite publisher's extra information. Only a `@type` and `@id`, which are correct (`Person` and `http://local.wordpress.test/#person`).

jdevalk commented:
> For pages when the knowledge graph is on a different person than the author of the page. We do not get to see the author information anywhere. Just to confirm, is this on purpose?
> On schema.org, the WebPage schema does have an author field.

Yes this is on purpose. Many themes don't output author info on pages, so we don't want to leak them.

> For author archive page when the knowledge graph is on a different person than the archive is for. We do not get to see the WebSite publisher's extra information. Only a @type and @id, which are correct (Person and http://local.wordpress.test/#person).

This is a bug, fixing.

igorschoester commented:
Acceptance ✅ 

Todo point `Update config wizard step.` is moved to this issue https://github.com/Yoast/wordpress-seo/issues/12516
